# Iniciando a exploração dos Dados


Até agora nós vimos python puro. Vamos lidar com ferramentas usadas no dia-a-dia para facilitar a análise de dados.

![pandas](https://static.boredpanda.com/blog/wp-content/uploads/2014/05/funny-animals-doing-yoga-20.jpg)

## PANDAS 
É a biblioteca mais popular para analisar dados. Com ela importar dado, filtrar e gerar gráficos é bem fácil.

## Ok?! E o que são bibliotecas?
É um conjunto de funções que as pessoas criam com a ideia de reaproveitamento de código.

## Como eu uso as bibliotecas? 
As bibliotecas precisam ser instaladas e "importadas" para o seu código. Nós já instalamos algumas bem legais nos nossos servidores e podemos só usá-las <3

## Open source <3
O Pandas é feito por uma comunidade de desenvolvedores de todo mundo, você pode fazer parte dessa comunidade, desenvolvendo, listando problemas (issues) e melhorando a documentação. Você ajuda acessando esse [GitHub](https://github.com/pandas-dev/pandas).



In [2]:
import pandas

### 1 - Vamos começar lendo um CSV (Comma-separated values) com o pandas

In [4]:
titanic = pandas.read_csv("https://databootcamp.nyc3.digitaloceanspaces.com/titanic.csv")

![titanic](http://thespiritscience.net/wp-content/uploads/2016/06/Titanic-sinking.jpg)

Assim como leitura de csv, tem vários tipos de leitura de arquivo.

**Dica: Quando você quiser saber oque é possível fazer com aquela função o Jupyter tem um atalho para isso que é o `?`**

In [5]:
pandas.read_csv?

**Dica: O Jupyter tem autocomplete, use bastante! Você acessa ele apertando o `tab`**

In [ ]:
#pandas.read_ # aperte o tab

No nosso caso o arquivo veio realmente separado por vírgula.

Se não viesse poderíamos usar o parâmetro `sep`

```
titanic = pandas.read_csv("titanic.csv", sep=",")
```

### 2 - Olhando algumas linhas do DataFrame

A função `head` retorna as primeiras 5 linhas do seu DataFrame.

In [6]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [7]:
type(titanic)

pandas.core.frame.DataFrame

#### DataFrame é um tipo do pandas que representa os dados de forma tabular, parecendo muito com uma tabela do Excel.
![](https://www.tutorialspoint.com/python_pandas/images/structure_table.jpg)

E se as 5 linhas não forem o suficiente?

In [8]:
titanic.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

**Por que usamos o Titanic?**

O Titanic é um dataset baseado em um dado público, ele ficou popular na comunidade de ciência de dados por ter bastante problemas conhecidos quando analisamos dados nas nossas empresas, dada as características e problemas desse dataset nós conseguimos simular cenários reais.

A sua popularidade ficou maior quando se tornou um desafio do Kaggle, uma dica que vai ajudar muito nas suas aulas é você analisar alguns notebooks já prontons na própria plataforma.

[Desafio do Titanic no Kaggle](https://www.kaggle.com/c/titanic/)

**O que significa cada coluna?**

* survived: Sobreviventes
    - (0 = Não; 1 = Sim)
* pclass: Classe do passageiro
     - (1 = Primeira; 2 = Segunda; 3 = Terceira)
* name: Nome
* sex: Sexo
* age: Idade
* sibsp: Número de irmão/cônjuges a bordo
* parch: Número de filhos/pais a bordo
* ticket: Número do ticket
* fare: Preço da Passagem
* cabin: Cabine
* embarked: Porto de embarque
    * [C = Cherbourg (França); Q = Queenstown (Irlanda); S = Southampton (Inglaterra)]

### Series
**Como retornamos os dados de uma única coluna?**

In [9]:
titanic["PassengerId"].head()  # parece um dicionário!

0    1
1    2
2    3
3    4
4    5
Name: PassengerId, dtype: int64

**Também é possível acessar por `.` "ponto"**

In [10]:
titanic.PassengerId.head()

0    1
1    2
2    3
3    4
4    5
Name: PassengerId, dtype: int64

**O tipo de uma coluna é chamado de Series**

In [11]:
type(titanic.PassengerId)

pandas.core.series.Series

Uma Série é uma lista, com uma dimensão, indexada.

![](https://i0.wp.com/www.emilkhatib.com/wp-content/uploads/2015/10/SeriesEN.png)

## Chegou a sua vez!

### **Desafio:** Selecione a coluna `Sex` e exiba os primeiros 20 valores
![alt text](https://static.vix.com/pt/sites/default/files/styles/large/public/a/atrasada-relogio-pulso-1116-1400x800.jpg?itok=qv3gUH6U)

In [16]:
#solução
titanic.Sex.head(20)

0       male
1     female
2     female
3     female
4       male
5       male
6       male
7       male
8     female
9     female
10    female
11    female
12      male
13      male
14    female
15    female
16      male
17      male
18    female
19    female
Name: Sex, dtype: object

**Quando precisamos retornar mais de uma coluna criamos uma lista**

In [17]:
col_id_sex = ["PassengerId","Sex"]

titanic[col_id_sex].head()

PassengerId     Sex
0            1    male
1            2  female
2            3  female
3            4  female
4            5    male

Quando retornamos mais de uma coluna ao invés de retornarmos uma `Serie` estamos retornando um `DataFrame`.

In [18]:
type(titanic[col_id_sex])

pandas.core.frame.DataFrame

**Dica: Crie uma lista separada e com nomes que façam sentido para evitar confusão ao longo da análise.**

Evite fazer tudo em uma linha só, isso dificulta a legibilidade.

In [19]:
titanic[["PassengerId","Sex"]].head()

PassengerId     Sex
0            1    male
1            2  female
2            3  female
3            4  female
4            5    male

### 3 - Algumas informações do DataFrame

**Qual a dimensão dos dados (Linhas x Colunas)?**

In [20]:
titanic.shape

(891, 12)

Ao importar o CSV, o Dataframe inclui uma coluna index, que não esta no arquivo. **Essa coluna é como se fosse um índice de linha do Excel.**

In [21]:
titanic.index

RangeIndex(start=0, stop=891, step=1)

**Quais são as colunas do Dataframe?**

In [22]:
titanic.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### **Desafio:** Retorne as primeiros 5 linhas de um DataFrame com todas as colunas menos a `Embarked`?

Lembrete: `Embarked` é a última coluna.
![alt text](https://static.vix.com/pt/sites/default/files/styles/large/public/a/atrasada-relogio-pulso-1116-1400x800.jpg?itok=qv3gUH6U)

In [52]:
#solução
titanic[titanic.columns[1:len(titanic.columns)-1]].head(5)

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin  
0    male  22.0      1      0         A/5 21171   7.2500   NaN  
1  female  38.0      1      0          PC 17599  71.2833   C85  
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN  
3  female  35.0      1      0            113803  53.1000  C123  
4    male  35.0      0      0            373450   8.0500   NaN

**Quais são os tipos de cada coluna?**


| Pandas dtype  | Python type  | Uso                                                   |
|---------------|--------------|-------------------------------------------------------|
| object        | str or mixed | Texto or mistura de valores numérico and não numérico |
| int64         | int          | Números inteiros                                      |
| float64       | float        | Números decimais (ponto flutuante)                    |
| bool          | bool         | Valores booleanos True/False                          |
| datetime64    | -            | Valores de data e hora                                |
| timedelta[ns] | -            | Diferença entre 2 dados de tempo                      |
| category      | -            | Lista finita de dados textuais                        |

In [53]:
titanic.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

**E se quisermos só as colunas de um certo tipo?**

In [54]:
titanic.dtypes == "int64"

PassengerId     True
Survived        True
Pclass          True
Name           False
Sex            False
Age            False
SibSp           True
Parch           True
Ticket         False
Fare           False
Cabin          False
Embarked       False
dtype: bool

Ele funciona como um operador lógico de `if`, ele vai em cada coluna e pergunta: _"Você é um int64?"_ a coluna vai responder se sim (`True`) ou não (`False`). 

A resposta desse filtro lógico é uma `Serie`.

In [56]:
filtro_tipo_numerico = (titanic.dtypes == "int64")

type(filtro_tipo_numerico)

pandas.core.series.Series

Não é necessário criar uma variável com o output desse filtro usando parenteses `()` mas dessa forma fica mais legível para com se confundir a igualdade `==` com a atribuição `=`.

**Vamos ao proposto, como filtramos só listas numéricas?**

In [57]:
titanic.select_dtypes(include=["int64"]).head()

PassengerId  Survived  Pclass  SibSp  Parch
0            1         0       3      1      0
1            2         1       1      1      0
2            3         1       3      0      0
3            4         1       1      1      0
4            5         0       3      0      0

### 4 - Informações estatísticas do DataFrame

| Informação Estatística | Categórico | Numérico | Object | Descrição                                           |
|------------------------|------------|----------|--------|-----------------------------------------------------|
| count                  | Sim        | Sim      | Sim    | Contagem que valores totais                         |
| unique                 | Sim        | Não      | Sim    | Contagem de elementos únicos                        |
| top                    | Sim        | Não      | SIm    | Valor que mais aparece                              |
| freq                   | Sim        | Não      | Sim    | Quantidade de vezes que valor mais aparece          |
| mean                   | Não        | Sim      | Não    | Média dos números                                   |
| std                    | Não        | Sim      | Não    | Desvio padrão de todos os números                   |
| min                    | Não        | Sim      | Não    | Valor mínimo dos números                            |
| 25%                    | Não        | Sim      | Não    | Valor que representa o primeiro quartil dos números |
| 50%                    | Não        | Sim      | Não    | Mediana dos números                                 |
| 75%                    | Não        | Sim      | Não    | Valor que representa o terceiro quartil dos números |
| max                    | Não        | Sim      | Não    | Valor máximo dos números                            |

**Estatísticas dos dados numéricos**

In [58]:
titanic.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

**Se quisermos acessar as estatísticas de somente uma coluna base selecionar ela**

In [59]:
titanic["Age"].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

**Nós não temos estatísticas somente dos dados numéricos mas os outros tipo de dados tem que se olhados com ainda mais cuidado**

In [60]:
titanic.describe(include="all")

PassengerId    Survived      Pclass                 Name   Sex  \
count    891.000000  891.000000  891.000000                  891   891   
unique          NaN         NaN         NaN                  891     2   
top             NaN         NaN         NaN  Ringhini, Mr. Sante  male   
freq            NaN         NaN         NaN                    1   577   
mean     446.000000    0.383838    2.308642                  NaN   NaN   
std      257.353842    0.486592    0.836071                  NaN   NaN   
min        1.000000    0.000000    1.000000                  NaN   NaN   
25%      223.500000    0.000000    2.000000                  NaN   NaN   
50%      446.000000    0.000000    3.000000                  NaN   NaN   
75%      668.500000    1.000000    3.000000                  NaN   NaN   
max      891.000000    1.000000    3.000000                  NaN   NaN   

               Age       SibSp       Parch    Ticket        Fare    Cabin  \
count   714.000000  891.000000  891.000000       891  891.000000      204   
unique         NaN         NaN         NaN       681         NaN      147   
top            NaN         NaN         NaN  CA. 2343         NaN  B96 B98   
freq           NaN         NaN         NaN         7         NaN        4   
mean     29.699118    0.523008    0.381594       NaN   32.204208      NaN   
std      14.526497    1.102743    0.806057       NaN   49.693429      NaN   
min       0.420000    0.000000    0.000000       NaN    0.000000      NaN   
25%      20.125000    0.000000    0.000000       NaN    7.910400      NaN   
50%      28.000000    0.000000    0.000000       NaN   14.454200      NaN   
75%      38.000000    1.000000    0.000000       NaN   31.000000      NaN   
max      80.000000    8.000000    6.000000       NaN  512.329200      NaN   

       Embarked  
count       889  
unique        3  
top           S  
freq        644  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

In [61]:
titanic['Embarked'].describe()

count     889
unique      3
top         S
freq      644
Name: Embarked, dtype: object

No `Embarked` o describe trás bastante valor, ele diz que 644 dos 889 passageiros embarcaram em `S` (Southampton).

Porém se olhar a coluna `Name` ela não funciona bem.

In [62]:
titanic['Name'].describe()

count                     891
unique                    891
top       Ringhini, Mr. Sante
freq                        1
Name: Name, dtype: object

### **Desafio:** Retorne a descrição apenas das colunas `Age` e `Pclass`.

![alt text](https://static.vix.com/pt/sites/default/files/styles/large/public/a/atrasada-relogio-pulso-1116-1400x800.jpg?itok=qv3gUH6U)


In [64]:
#solução
titanic[['Age', 'Pclass']].describe()

Age      Pclass
count  714.000000  891.000000
mean    29.699118    2.308642
std     14.526497    0.836071
min      0.420000    1.000000
25%     20.125000    2.000000
50%     28.000000    3.000000
75%     38.000000    3.000000
max     80.000000    3.000000

**Nós conseguimos acessar aos dados do describe diretamente**

In [65]:
titanic["Age"].mean()

29.69911764705882

In [66]:
titanic["Age"].std()

14.526497332334044